### PR TITLE
Use entrypoint instead of command to capture kill signals.

### PR DESCRIPTION
**Pull request by bmoyles0117**

### CODE REVIEW

File: docker/entrypoint.sh [hunk 1]
```diff
@@ -9,4 +11,23 @@ if [ -n "$METADATA_AGENT_URL" ]; then
...
+if [ "${1:0:1}" = '-' ]; then
+  /usr/sbin/google-fluentd "$@" &
+else
+  exec "$@" &
+fi
+
+pid=$!
+wait $pid
+exit_code="$?"
+wait $pid
```
Comment (igorpeshansky) [inline docker/entrypoint.sh]: Can you please add a comment explaining the double wait?

File: docker/entrypoint.sh [hunk 2]
```diff
@@ -9,4 +11,23 @@ if [ -n "$METADATA_AGENT_URL" ]; then
...
+
+# Ignore 143, as that is the code when SIGINT kills the process.
+if [[ exit_code -eq "143" ]]; then
```
Comment (igorpeshansky) [inline docker/entrypoint.sh]: Did you mean `$exit_code`?

File: docker/entrypoint.sh [hunk 3]
```diff
@@ -9,4 +11,23 @@ if [ -n "$METADATA_AGENT_URL" ]; then
...
+else
+  exec "$@" &
+fi
+
+pid=$!
+wait $pid
+exit_code="$?"
+wait $pid
+
+# Ignore 143, as that is the code when SIGINT kills the process.
```
Comment (igorpeshansky) [inline docker/entrypoint.sh]: I think you meant SIGTERM. SIGINT is 130.

File: docker/entrypoint.sh [hunk 4]
```diff
@@ -9,4 +11,23 @@ if [ -n "$METADATA_AGENT_URL" ]; then
...
 fi
 
-/usr/sbin/google-fluentd "$@"
+trap 'kill $pid' TERM INT
+
+# first arg is a flag.
+if [ "${1:0:1}" = '-' ]; then
+  /usr/sbin/google-fluentd "$@" &
+else
+  exec "$@" &
```
Comment (igorpeshansky) [inline docker/entrypoint.sh]: Why is this needed?

File: docker/entrypoint.sh [hunk 5]
```diff
@@ -9,4 +11,23 @@ if [ -n "$METADATA_AGENT_URL" ]; then
     /etc/google-fluentd/google-fluentd.conf
 fi
 
-/usr/sbin/google-fluentd "$@"
+trap 'kill $pid' TERM INT
+
+# first arg is a flag.
```
Comment (qingling128) [inline docker/entrypoint.sh]: `s/first/First`.

File: docker/entrypoint.sh (hunk 4 above)
Comment (qingling128) [reply docker/entrypoint.sh]: Same question.

File: docker/entrypoint.sh (hunk 3 above)
Comment (bmoyles0117) [reply docker/entrypoint.sh]: SIGINT is the default kill signal sent by `docker stop`, that's the code I'm referring to, probably should make that clearer.

File: docker/entrypoint.sh [hunk 6]
```diff
@@ -9,4 +11,23 @@ if [ -n "$METADATA_AGENT_URL" ]; then
...
+wait $pid
+exit_code="$?"
+wait $pid
+
+# Ignore 143, as that is the code when SIGINT kills the process.
+if [[ exit_code -eq "143" ]]; then
+  exit 0
+fi
+
+exit $exit_code
```
Comment (qingling128) [inline docker/entrypoint.sh]: What are those two characters at the end of line? Just some weird display issue with Github?

Comment (bmoyles0117) [reply docker/entrypoint.sh]: I think that's saying that there's no new line at the end of the file.

File: docker/entrypoint.sh (hunk 4 above)
Comment (bmoyles0117) [reply docker/entrypoint.sh]: This is a typical thing to see in docker entrypoints. This allows a user to run their own commands, such as "bash".

If the command looks something like
```
docker run -d logging-agent -o /var/log/google-fluentd.log
```
it will use the first condition, passing the flags to the google-fluentd process.

If the command looks like 
```
docker run -d logging-agent bash
```

It will exec "bash".

If the command looks like 
```
docker run -d logging-agent
```

It will exec the CMD defined in the dockerfile, which is google-fluentd.

Comment (qingling128) [reply docker/entrypoint.sh]: Makes sense. Thanks for explaining. Do you think it's worth commenting it with a brief clarification? Just for the benefit of people who are newbies for the Docker world.

File: docker/entrypoint.sh (hunk 3 above)
Comment (qingling128) [reply docker/entrypoint.sh]: Yeah, maybe rephrase the comment a bit to make that clear.

File: docker/entrypoint.sh (hunk 6 above)
Comment (qingling128) [reply docker/entrypoint.sh]: Ah, yeah... Apparently. Just hovered on it and saw the explanation. LOL

File: docker/entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+# docker command. When a flag is sent, it will be passed on to the fluentd
+# process. Anything else will be interpreted as a command to be run.
+#
+# Passing a flag.
+# $ docker run -it {image:tag} -o /var/log/google-fluentd.log
+#
+# Passing a command.
+# $ docker run -it {image:tag} /bin/bash
+#
+# Default behavior uses CMD defined in Dockerfile.
```
Comment (igorpeshansky) [inline docker/entrypoint.sh]: What's counterintuitive is that the CMD is passed into the `else` branch below... But I don't think it matters much to the user.

File: docker/entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+if [ -z "${METADATA_AGENT_URL:-}" -a -n "${METADATA_AGENT_HOSTNAME:-}" ]; then
+  METADATA_AGENT_URL="http://${METADATA_AGENT_HOSTNAME}:8000"
+fi
+if [ -n "$METADATA_AGENT_URL" ]; then
+  sed -i "s,http://local-metadata-agent.stackdriver.com:8000,$METADATA_AGENT_URL," \
+    /etc/google-fluentd/google-fluentd.conf
+fi
+
+# This docker image supports sending either a flag, or a command, as the
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply docker/entrypoint.sh]: Done.

File: docker/entrypoint.sh [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  METADATA_AGENT_URL="http://${METADATA_AGENT_HOSTNAME}:8000"
+fi
+if [ -n "$METADATA_AGENT_URL" ]; then
+  sed -i "s,http://local-metadata-agent.stackdriver.com:8000,$METADATA_AGENT_URL," \
+    /etc/google-fluentd/google-fluentd.conf
+fi
+
+# This docker image supports sending either a flag, or a command, as the 
+# docker command. When a flag is sent, it will be passed on to the fluentd
+# process. Anything else will be interpreted as a command to be run.
```
[the remark this answers is not in the document]
Comment (bmoyles0117) [reply docker/entrypoint.sh]: Done.

File: docker/entrypoint.sh (hunk 7 above)
Comment (bmoyles0117) [reply docker/entrypoint.sh]: Are you suggesting to remove this final example?

Comment (igorpeshansky) [reply docker/entrypoint.sh]: No, just musing...
